aggregator_core/src/datastore.rs

```diff
@@ -3468,24 +3468,53 @@ impl<C: Clock> Transaction<'_, C> {
         task_id: &TaskId,
         batch_id: &BatchId,
     ) -> Result<(), Error> {
-        // XXX
+        // XXX: update tests
         let stmt = self
             .prepare_cached(
                 "INSERT INTO outstanding_batches (task_id, batch_id)
-                VALUES ((SELECT id FROM tasks WHERE task_id = $1), $2)",
+                VALUES ((SELECT id FROM tasks WHERE task_id = $1), $2)
+                ON CONFLICT DO NOTHING
+                RETURNING UPPER((SELECT client_timestamp_interval FROM batches WHERE batch_identifier = $2)) < COALESCE($3::TIMESTAMP - (SELECT report_expiry_age FROM tasks WHERE task_id = $1) * '1 second'::INTERVAL, '-infinity'::TIMESTAMP) AS is_expired",
+            )
+            .await?;
+        let rows = self
+            .query(
+                &stmt,
+                &[
+                    /* task_id */ task_id.as_ref(),
+                    /* batch_id */ batch_id.as_ref(),
+                    /* now */ &self.clock.now().as_naive_date_time()?,
+                ],
             )
             .await?;
+        let row_count = rows.len().try_into()?;
 
-        self.execute(
-            &stmt,
-            &[
-                /* task_id */ task_id.as_ref(),
-                /* batch_id */ batch_id.as_ref(),
-            ],
-        )
-        .await?;
+        if let Some(row) = rows.into_iter().next() {
+            // We got a row back, meaning we wrote an outstanding batch. We check that it wasn't
+            // expired per the task's report_expiry_age, but otherwise we are done.
+            if row.get("is_expired") {
```

Comment:
Similar to my point above, can we check for expiry outside of sql? Well... we can't outside of sql because we have to select the batch's client_timestamp_interval.

So in this case I don't think the same feedback applies. It's less expensive to unconditionally insert and rarely delete versus unconditionally select then insert. (sorry, just thinking out loud).

Reply:
Yup, that's the right reasoning. See my response to your other comment along these lines for how I think about this approach.